src/Task.cc

```diff
@@ -430,8 +451,11 @@ void Task::on_syscall_exit_arch(int syscallno, const Registers& regs) {
       if (amount > 0) {
         ranges.push_back(FileMonitor::Range(regs.arg2(), amount));
       }
-      fd_table()->did_write(this, fd, ranges,
-                            get_io_offset_arch<Arch>(syscallno, regs));
+      int64_t offset = -1;
+      if (fd_table()->needs_offset(this, fd, true)) {
+        offset = get_io_offset_arch<Arch>(this, syscallno, regs);
+      }
+      fd_table()->did_write(this, fd, ranges, offset);
```

Comment:
This `needs_offset` boilerplate is a bit unfortunate. Would it not be better to pass a couple of extra parameters (`Task*` and `fd`) to `did_write` etc so the `FileMonitor` can find the offset itself if needed?